### ISSUE TITLE
Segfault while loading remote WASM module

### ISSUE THREAD

**Issue by mandarjog** [envoyproxy/envoy-wasm#424]
1.5 latest proxy build from 2020-02-20

I have a wasm module that loads successfully if loaded as a local file and segfaults if loaded remotely.

```yaml
         - name: envoy.filters.http.wasm
            typed_config:
              "@type": "type.googleapis.com/udpa.type.v1.TypedStruct"
              type_url: type.googleapis.com/envoy.config.filter.http.wasm.v2.Wasm
              value:
                config:
                  vm_config:
                    runtime: envoy.wasm.runtime.v8
                    code:
                      local1:
                        #inline_string: "envoy.wasm.jwt_header"
                        filename: /Volumes/work/go/src/istio.io/proxy15/extensions/jwt_header/jwt_header_plugin_15.wasm
                      remote:
                        sha256: 497302bb0de35537b066118bd9bfcd893fe46f04b32005a114ea0d2fc3683fab
                        http_uri:
                          uri: http://storage.googleapis.com/artifacts.mixologist-142215.appspot.com/extensions/jwt_header_plugin_15.wasm
                          cluster: googleapis_storage
                          timeout:
                            seconds: 5
                  configuration: |
                    { "header_map": {"jwt-group": "group"} }
```

Debug log
```
[2020-02-23 21:55:07.240][9581077][debug][init] [external/envoy/source/common/init/manager_impl.cc:45] init manager Cluster googleapis_storage contains no targets
[2020-02-23 21:55:07.240][9581077][debug][init] [external/envoy/source/common/init/watcher_impl.cc:14] init manager Cluster googleapis_storage initialized, notifying ClusterImplBase
[2020-02-23 21:55:07.240][9581077][debug][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:1084] membership update for TLS cluster googleapis_storage added 1 removed 0
[2020-02-23 21:55:07.240][9581077][debug][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:106] cm init: init complete: cluster=googleapis_storage primary=0 secondary=0
[2020-02-23 21:55:07.240][9581077][debug][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:126] maybe finish initialize state: 1
[2020-02-23 21:55:07.240][9581077][debug][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:134] maybe finish initialize primary init clusters empty: true
[2020-02-23 21:55:07.240][9581077][debug][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:142] maybe finish initialize secondary init clusters empty: true
[2020-02-23 21:55:07.240][9581077][debug][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:165] maybe finish initialize cds api ready: false
[2020-02-23 21:55:07.240][9581077][info][upstream] [external/envoy/source/common/upstream/cluster_manager_impl.cc:171] cm init: all clusters initialized
[2020-02-23 21:55:07.240][9581077][info][main] [external/envoy/source/server/server.cc:529] all clusters initialized. initializing init manager
[2020-02-23 21:55:07.240][9581077][debug][init] [external/envoy/source/common/init/manager_impl.cc:49] init manager Server initializing
[2020-02-23 21:55:07.240][9581077][debug][init] [external/envoy/source/common/init/target_impl.cc:15] init manager Server initializing target RemoteAsyncDataProvider
[2020-02-23 21:55:07.240][9581077][debug][config] [external/envoy/source/common/config/remote_data_fetcher.cc:35] fetch remote data from [uri = http://storage.googleapis.com/artifacts.mixologist-142215.appspot.com/extensions/jwt_header_plugin_15.wasm]: start
[2020-02-23 21:55:07.241][9581077][debug][router] [external/envoy/source/common/router/router.cc:474] [C0][S15007244551492852564] cluster 'googleapis_storage' match for URL '/artifacts.mixologist-142215.appspot.com/extensions/jwt_header_plugin_15.wasm'
[2020-02-23 21:55:07.241][9581077][debug][router] [external/envoy/source/common/router/router.cc:614] [C0][S15007244551492852564] router decoding headers:
':path', '/artifacts.mixologist-142215.appspot.com/extensions/jwt_header_plugin_15.wasm'
':authority', 'storage.googleapis.com'
':method', 'GET'
':scheme', 'http'
'x-envoy-internal', 'true'
'x-forwarded-for', '192.168.86.189'
'x-envoy-expected-rq-timeout-ms', '5000'

[2020-02-23 21:55:07.241][9581077][debug][pool] [external/envoy/source/common/http/http1/conn_pool.cc:95] creating a new connection
[2020-02-23 21:55:07.241][9581077][debug][client] [external/envoy/source/common/http/codec_client.cc:34] [C0] connecting
[2020-02-23 21:55:07.241][9581077][debug][connection] [external/envoy/source/common/network/connection_impl.cc:691] [C0] connecting to 172.217.5.112:80
[2020-02-23 21:55:07.241][9581077][debug][connection] [external/envoy/source/common/network/connection_impl.cc:700] [C0] connection in progress
[2020-02-23 21:55:07.241][9581077][debug][pool] [external/envoy/source/common/http/conn_pool_base.cc:55] queueing request due to no available connections
[2020-02-23 21:55:07.256][9581077][debug][connection] [external/envoy/source/common/network/connection_impl.cc:563] [C0] connected
[2020-02-23 21:55:07.256][9581077][debug][client] [external/envoy/source/common/http/codec_client.cc:72] [C0] connected
[2020-02-23 21:55:07.256][9581077][debug][pool] [external/envoy/source/common/http/http1/conn_pool.cc:244] [C0] attaching to next request
[2020-02-23 21:55:07.256][9581077][debug][router] [external/envoy/source/common/router/router.cc:1711] [C0][S15007244551492852564] pool ready
[2020-02-23 21:55:07.279][9581077][debug][router] [external/envoy/source/common/router/router.cc:1115] [C0][S15007244551492852564] upstream headers complete: end_stream=false
[2020-02-23 21:55:07.281][9581077][debug][http] [external/envoy/source/common/http/async_client_impl.cc:95] async http request response headers (end_stream=false):
':status', '200'
'x-guploader-uploadid', 'AEnB2UovJkd7_J4o4zkejGOnSWuJK0nOYGM_jE7r3gmAtx1Uz1LWQwBOrwAbBXZzzP_9GI1dY4kF-u9fjdIYXevhHnAFyWH6Ag'
'expires', 'Mon, 24 Feb 2020 06:45:43 GMT'
'date', 'Mon, 24 Feb 2020 05:45:43 GMT'
'last-modified', 'Mon, 24 Feb 2020 05:28:12 GMT'
'etag', '"795b2271cce03e54af2102becf63f921"'
'x-goog-generation', '1582522092452966'
'x-goog-metageneration', '1'
'x-goog-stored-content-encoding', 'identity'
'x-goog-stored-content-length', '1007490'
'content-type', 'application/wasm'
'x-goog-hash', 'crc32c=hkiPJg=='
'x-goog-hash', 'md5=eVsicczgPlSvIQK+z2P5IQ=='
'x-goog-storage-class', 'STANDARD'
'accept-ranges', 'bytes'
'content-length', '1007490'
'server', 'UploadServer'
'cache-control', 'public, max-age=3600'
'age', '564'
'x-envoy-upstream-service-time', '37'

[2020-02-23 21:55:07.405][9581077][debug][client] [external/envoy/source/common/http/codec_client.cc:104] [C0] response complete
[2020-02-23 21:55:07.407][9581077][debug][config] [external/envoy/source/common/config/remote_data_fetcher.cc:45] fetch remote data [uri = http://storage.googleapis.com/artifacts.mixologist-142215.appspot.com/extensions/jwt_header_plugin_15.wasm]: success
[2020-02-23 21:55:07.497][9581077][info][wasm] [external/envoy/source/extensions/common/wasm/context.cc:521] wasm log getProperty plugin_root_id
[2020-02-23 21:55:07.498][9581077][info][wasm] [external/envoy/source/extensions/common/wasm/context.cc:539] wasm log getProperty plugin_root_id part=plugin_root_id
[2020-02-23 21:55:07.498][9581077][info][wasm] [external/envoy/source/extensions/common/wasm/context.cc:408] wasm log FindValue plugin_root_id
[2020-02-23 21:55:07.498][9581077][info][wasm] [external/envoy/source/extensions/common/wasm/context.cc:549] wasm log getProperty plugin_root_id part=plugin_root_id
[2020-02-23 21:55:07.498][9581077][critical][backtrace] [bazel-out/darwin-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:83] Caught Segmentation fault: 11, suspect faulting address 0x5bb
[2020-02-23 21:55:07.498][9581077][critical][backtrace] [bazel-out/darwin-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:70] Backtrace (use tools/stack_decode.py to get line numbers):
[2020-02-23 21:55:07.498][9581077][critical][backtrace] [bazel-out/darwin-dbg/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71] Envoy version: a41e0fc4c6394650f0b42030573f778648e42152/1.13.1-dev/Modified/DEBUG/BoringSSL
AsyncClient 0x296bb10, stream_id_: 15007244551492852564
&stream_info_:
  StreamInfoImpl 0x296bcb0, protocol_: 1, response_code_: 200, response_code_details_: via_upstream, health_check_request_: 0, route_name_:
./run_test.sh: line 23: 63830 Segmentation fault: 11  ${BAZEL_BIN}/src/envoy/envoy -c ${WD}/testdata/server.yaml --concurrency 2 --allow-unknown-fields -l debug
```

**Comment by mandarjog** [comment]
@jplevyak @PiotrSikora @kyessenov

**Comment by duderino** [comment]
@yxue wrote the impl I think?  or was it @lambdai?  or both?

**Comment by jplevyak** [comment]
@yxue did the PR for proxy-wasm.

**Comment by jplevyak** [comment]
@mandarjog could you post your entire config YAML file ?

**Comment by mandarjog** [comment]
@jplevyak https://github.com/istio/proxy/blob/a710ade894cd315a4c81d50c985b33bdd30d2150/extensions/jwt_header/testdata/server.yaml

**Comment by yxue** [comment]
Looking

**Comment by mandarjog** [comment]
Also found that it does not matter if I try loading a different module, still I get the same segfault.
In my 2nd experiment, I forgot to update sha256 of the other module, but still the same error.

**Comment by yxue** [comment]
The same error means the same stack trace? If the sha256 is incorrect, the error is expected since it will throw an exception.

**Comment by jplevyak** [comment]
Here is what I found out.  The error is coming from line wasm_filter.cc:46

auto configuratiauto configuration = std::make_shared<std::string>(config.config().configuration();

The specific problem is with the final .configuration() which returns 0x1:

(gdb) p config.config()
$7 = (const envoy::extensions::wasm::v3::PluginConfig &) @0x55555c341200: warning: RTTI symbol not found for class 'std::_Sp_counted_ptr_inplace<absl::flat_hash_map<unsigned short, std::shared_ptr<Envoy::Network::FilterChain>, absl::hash_internal::Hash<unsigned short>, std::equal_to<unsigned short>, std::allocator<std::pair<unsigned short const, std::shared_ptr<Envoy::Network::FilterChain> > > >, std::allocator<absl::flat_hash_map<unsigned short, std::shared_ptr<Envoy::Network::FilterChain>, absl::hash_internal::Hash<unsigned short>, std::equal_to<unsigned short>, std::allocator<std::pair<unsigned short const, std::shared_ptr<Envoy::Network::FilterChain> > > > >, (__gnu_cxx::_Lock_policy)2>'

{<google::protobuf::Message> = {<google::protobuf::MessageLite> = {
      _vptr$MessageLite = 0x55555b4841e8 <vtable for std::_Sp_counted_ptr_inplace<absl::flat_hash_map<unsigned short, std::shared_ptr<Envoy::Network::FilterChain>, absl::hash_internal::Hash<unsigned short>, std::equal_to<unsigned short>, std::allocator<std::pair<unsigned short const, std::shared_ptr<Envoy::Network::FilterChain> > > >, std::allocator<absl::flat_hash_map<unsigned short, std::shared_ptr<Envoy::Network::FilterChain>, absl::hash_internal::Hash<unsigned short>, std::equal_to<unsigned short>, std::allocator<std::pair<unsigned short const, std::shared_ptr<Envoy::Network::FilterChain> > > > >, (__gnu_cxx::_Lock_policy)2>+16>}, <No data fields>},
  static kIndexInFileMessages = 1,
  _internal_metadata_ = {<google::protobuf::internal::InternalMetadataWithArenaBase<google::protobuf::UnknownFieldSet, google::protobuf::internal::InternalMetadataWithArena>> = {ptr_ = 0x100000003, static kPtrTagMask = 1,
      static kPtrValueMask = -2}, <No data fields>}, name_ = {
    ptr_ = 0x55555c354150}, root_id_ = {ptr_ = 0x55555c354168}, configuration_ = {
    ptr_ = 0x1}, vm_config_ = 0x1, _cached_size_ = {
    size_ = {<std::__atomic_base<int>> = {static _S_alignment = 4,
        _M_i = 0}, <No data fields>}}}

This is because the callback is occurring after "config" has been deleted.  

We need to either pass config by value or make a shared_ptr copy of it and pass the shared pointer to the callback.

**Comment by jplevyak** [comment]
@yxue  Do you want me to make the fix ?

**Comment by mandarjog** [comment]
Thanks @jplevyak it will have to go back into 1.5

**Comment by mandarjog** [comment]
Config by value seems like the right thing no?
The http response was deleted and then onConfigure was called ?

**Comment by jplevyak** [comment]
This is during VM startup.  The problem is that the callback for loading the VM from the code is async when the code is remote and the config is owned up the stack which is going away before the callback.

I'll make the PR

**Comment by jplevyak** [comment]
https://github.com/envoyproxy/envoy-wasm/pull/427

**Comment by LSChyi** [comment]
Is this issue fixed in 1.5.1? I tested with 1.5.1 and got the following message:

```
Envoy (Epoch 0)] [2020-04-23 06:02:22.585[24[critical[main [external/envoy/source/exe/terminate_handler.cc:13] std::terminate called! (possible uncaught exception, see trace)
[Envoy (Epoch 0)] [2020-04-23 06:02:22.585[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:70] Backtrace (use tools/stack_decode.py to get line numbers):
[Envoy (Epoch 0)] [2020-04-23 06:02:22.585[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71] Envoy version: 6cdd2c56746ef582c8a851ed234c71719f9e3742/1.13.1-dev/Clean/RELEASE/BoringSSL
[Envoy (Epoch 0)] [2020-04-23 06:02:22.593[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #0: Envoy::TerminateHandler::logOnTerminate()::$_0::operator()() [0x55afe25843ae
[Envoy (Epoch 0)] [2020-04-23 06:02:22.600[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:77] #1: [0x55afe25842b9
[Envoy (Epoch 0)] [2020-04-23 06:02:22.607[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #2: std::__terminate() [0x55afe2ae7073
[Envoy (Epoch 0)] [2020-04-23 06:02:22.615[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #3: Envoy::Config::DataSource::RemoteAsyncDataProvider::onFailure() [0x55afe13a9139
[Envoy (Epoch 0)] [2020-04-23 06:02:22.623[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #4: Envoy::Config::DataFetcher::RemoteDataFetcher::onSuccess() [0x55afe213cd1a
[Envoy (Epoch 0)] [2020-04-23 06:02:22.630[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #5: Envoy::Http::AsyncRequestImpl::onComplete() [0x55afe22a710c
[Envoy (Epoch 0)] [2020-04-23 06:02:22.637[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #6: Envoy::Http::AsyncStreamImpl::encodeData() [0x55afe22a6ad2
[Envoy (Epoch 0)] [2020-04-23 06:02:22.645[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #7: Envoy::Http::Utility::sendLocalReply() [0x55afe23bb62a
[Envoy (Epoch 0)] [2020-04-23 06:02:22.652[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #8: Envoy::Http::AsyncStreamImpl::sendLocalReply() [0x55afe22a7db5
[Envoy (Epoch 0)] [2020-04-23 06:02:22.659[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #9: Envoy::Router::Filter::onUpstreamAbort() [0x55afe22b1821
[Envoy (Epoch 0)] [2020-04-23 06:02:22.666[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #10: Envoy::Router::Filter::onUpstreamReset() [0x55afe22b1e4d
[Envoy (Epoch 0)] [2020-04-23 06:02:22.674[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #11: Envoy::Router::Filter::UpstreamRequest::onResetStream() [0x55afe22b53d3
[Envoy (Epoch 0)] [2020-04-23 06:02:22.681[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #12: Envoy::Http::Http2::ConnectionImpl::StreamImpl::resetStream() [0x55afe2325ac2
[Envoy (Epoch 0)] [2020-04-23 06:02:22.688[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #13: Envoy::Http::CodecClient::onEvent() [0x55afe22956ab
[Envoy (Epoch 0)] [2020-04-23 06:02:22.696[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #14: Envoy::Network::ConnectionImplBase::raiseConnectionEvent() [0x55afe20c602d
[Envoy (Epoch 0)] [2020-04-23 06:02:22.703[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #15: Envoy::Network::ConnectionImpl::raiseEvent() [0x55afe20c2212
[Envoy (Epoch 0)] [2020-04-23 06:02:22.710[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #16: Envoy::Network::ConnectionImpl::closeSocket() [0x55afe20c1cd8
[Envoy (Epoch 0)] [2020-04-23 06:02:22.718[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #17: Envoy::Network::ConnectionImpl::close() [0x55afe20c1977
[Envoy (Epoch 0)] [2020-04-23 06:02:22.725[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #18: Envoy::Http::CodecClient::onData() [0x55afe2295b25
[Envoy (Epoch 0)] [2020-04-23 06:02:22.732[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #19: Envoy::Http::CodecClient::CodecReadFilter::onData() [0x55afe229679d
[Envoy (Epoch 0)] [2020-04-23 06:02:22.739[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #20: Envoy::Network::FilterManagerImpl::onContinueReading() [0x55afe20c6b7b
[Envoy (Epoch 0)] [2020-04-23 06:02:22.747[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #21: Envoy::Network::ConnectionImpl::onReadReady() [0x55afe20c35ea
[Envoy (Epoch 0)] [2020-04-23 06:02:22.754[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #22: Envoy::Network::ConnectionImpl::onFileEvent() [0x55afe20c30bd
[Envoy (Epoch 0)] [2020-04-23 06:02:22.762[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #23: Envoy::Event::FileEventImpl::assignEvents()::$_0::__invoke() [0x55afe20bdb90
[Envoy (Epoch 0)] [2020-04-23 06:02:22.769[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #24: event_process_active_single_queue [0x55afe23eef8b
[Envoy (Epoch 0)] [2020-04-23 06:02:22.776[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #25: event_base_loop [0x55afe23ed81e
[Envoy (Epoch 0)] [2020-04-23 06:02:22.783[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #26: Envoy::Server::InstanceImpl::run() [0x55afe205be43
[Envoy (Epoch 0)] [2020-04-23 06:02:22.791[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #27: Envoy::MainCommonBase::run() [0x55afe0d37408
[Envoy (Epoch 0)] [2020-04-23 06:02:22.798[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #28: main [0x55afe0d36322
[Envoy (Epoch 0)] [2020-04-23 06:02:22.798[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #29: __libc_start_main [0x7f5bbd25fb97
[Envoy (Epoch 0)] [2020-04-23 06:02:22.798[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:83] Caught Aborted, suspect faulting address 0x53900000018
[Envoy (Epoch 0)] [2020-04-23 06:02:22.798[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:70] Backtrace (use tools/stack_decode.py to get line numbers):
[Envoy (Epoch 0)] [2020-04-23 06:02:22.798[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:71] Envoy version: 6cdd2c56746ef582c8a851ed234c71719f9e3742/1.13.1-dev/Clean/RELEASE/BoringSSL
[Envoy (Epoch 0)] [2020-04-23 06:02:22.798[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #0: __restore_rt [0x7f5bbd641890
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:77] #1: [0x55afe25842b9
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #2: std::__terminate() [0x55afe2ae7073
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #3: Envoy::Config::DataSource::RemoteAsyncDataProvider::onFailure() [0x55afe13a9139
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #4: Envoy::Config::DataFetcher::RemoteDataFetcher::onSuccess() [0x55afe213cd1a
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #5: Envoy::Http::AsyncRequestImpl::onComplete() [0x55afe22a710c
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #6: Envoy::Http::AsyncStreamImpl::encodeData() [0x55afe22a6ad2
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #7: Envoy::Http::Utility::sendLocalReply() [0x55afe23bb62a
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #8: Envoy::Http::AsyncStreamImpl::sendLocalReply() [0x55afe22a7db5
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #9: Envoy::Router::Filter::onUpstreamAbort() [0x55afe22b1821
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #10: Envoy::Router::Filter::onUpstreamReset() [0x55afe22b1e4d
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #11: Envoy::Router::Filter::UpstreamRequest::onResetStream() [0x55afe22b53d3
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #12: Envoy::Http::Http2::ConnectionImpl::StreamImpl::resetStream() [0x55afe2325ac2
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #13: Envoy::Http::CodecClient::onEvent() [0x55afe22956ab
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #14: Envoy::Network::ConnectionImplBase::raiseConnectionEvent() [0x55afe20c602d
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #15: Envoy::Network::ConnectionImpl::raiseEvent() [0x55afe20c2212
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #16: Envoy::Network::ConnectionImpl::closeSocket() [0x55afe20c1cd8
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #17: Envoy::Network::ConnectionImpl::close() [0x55afe20c1977
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #18: Envoy::Http::CodecClient::onData() [0x55afe2295b25
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #19: Envoy::Http::CodecClient::CodecReadFilter::onData() [0x55afe229679d
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #20: Envoy::Network::FilterManagerImpl::onContinueReading() [0x55afe20c6b7b
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #21: Envoy::Network::ConnectionImpl::onReadReady() [0x55afe20c35ea
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #22: Envoy::Network::ConnectionImpl::onFileEvent() [0x55afe20c30bd
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #23: Envoy::Event::FileEventImpl::assignEvents()::$_0::__invoke() [0x55afe20bdb90
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #24: event_process_active_single_queue [0x55afe23eef8b
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #25: event_base_loop [0x55afe23ed81e
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #26: Envoy::Server::InstanceImpl::run() [0x55afe205be43
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #27: Envoy::MainCommonBase::run() [0x55afe0d37408
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #28: main [0x55afe0d36322
[Envoy (Epoch 0)] [2020-04-23 06:02:22.805[24[critical[backtrace [bazel-out/k8-opt/bin/external/envoy/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:75] #29: __libc_start_main [0x7f5bbd25fb97
AsyncClient 0x55afe5ffc490, stream_id_: 3857136155226626854
&stream_info_:
  StreamInfoImpl 0x55afe5ffc630, protocol_: 1, response_code_: 503, response_code_details_: upstream_reset_before_response_started{connection termination}, health_check_request_: 0, route_name_:
2020-04-23T06:02:22.809132Z    info    sds    node:sidecar~10.1.4.251~productpage-v1-76589d9fdc-7p5zg.bookinfo~bookinfo.svc.cluster.local-1 resource:ROOTCA connection is terminated: rpc error: code = Canceled desc = context canceled
2020-04-23T06:02:22.809169Z    error    sds    Remote side closed connection
2020-04-23T06:02:22.809221Z    info    sds    node:sidecar~10.1.4.251~productpage-v1-76589d9fdc-7p5zg.bookinfo~bookinfo.svc.cluster.local-2 resource:default connection is terminated: rpc error: code = Canceled desc = context canceled
2020-04-23T06:02:22.809421Z    error    sds    Remote side closed connection
2020-04-23T06:02:22.809424Z    info    transport: loopyWriter.run returning. connection error: desc = "transport is closing"
2020-04-23T06:02:22.809475Z    error    Epoch 0 exited with error: signal: aborted
2020-04-23T06:02:22.809505Z    info    No more active epochs, terminating
```
